OpenEMS/openems
# FEMS Backports 2024-07

## Common/General
- Gradle: automate version update for android build.gradle
- `Controller.Ess.Limiter14a`: Edge Controller and UI visualization for "dimming" according to German law §14a EnWG

## OpenEMS Backend
- Performance optimizations
  - Drop PerMessageDeflate because it has bad effects on performance with a high number of connected Edges
  - Add a copy of Draft6455 which does not throw an exception internally
  - Reduce Exceptions
  - OnOpen: no unnecessary parsing of handshake + do not throw Exception
  - Parse JsonrpcMessage inside Executor Thread
  - sendMessage by default never throws Exception
  - use StringBuilders where feasible
  - throw the same Exception for all disposed requests in WsData
- InfluxDB: remove filter for sunspec channels
- Metadata: add handling of "ems type"
- Aggregated Influx: add channels for ctrlEssLimiter14a0
- Cleanup logs:
  - "No measurement provided for zoneId"
  - "Channels missing in Current-Data "
- WsData:
  - handle failed send message
  - set WsData.WebSocket via constructor
- Change debugLog interval to 5 seconds
- Replace run() with standard method names
- Add internal WebsocketPerformanceTestApp
- Drop now unused Websocket DebugMode

## OpenEMS Edge
- Home Battery: 
  - fix calculation of number of towers
  - handle tower exceptionally not 0/null for some bms versions
  - downgrade uncritical states to INFO
- Battery-Protection: add SoC-to-Percent characteristics & implement for FENECON Home Battery
- KACO Battery-Inverter: optimize state machine handling
- MR Gridcon: move obsolete bundle to archive
  - See https://github.com/fenecon/openems-archive/tree/main/io.openems.edge.ess.mr.gridcon
- Translation keys for timedata category

## OpenEMS UI
- Fix Nullpointer thrown through chart.js after failed query.
  - Apparently chartjs not having completely rendered datasets and property borderColor was missing
- Extend vscode extensions with 18n ally
  - add `i18n ally` to vscode extensions
- Fix ChartJs yAxis scaling for only 0 values in data series
  - Chartjs y-axis Scaling for only null or/and 0 values in data series results in doing negative and positive scaling
- Add missing redirect after token expired
  - Token expired, old route got reloaded, no redirection to `/login`
- Add Power-Assistant
  - Add Power-Assistant to UI Settings
  - Show relevant channels for power constraints; sorted by Scheduler
  - Cleanup Settings + separate ADMIN widgets
- Add subnavigation to consumption history
  - restructured overdue shared folder
  - added current and voltage chart (only visible for Role.INSTALLER to consumption and production history
  - added subnavigation to consumption history
- Add missing module for angular date picker
  - Add missing import for pickdatepopovercomponent, unittest will follow soon
- Fix live modal visualization of meters
- Add Unittest for [angular-my-date-picker](https://github.com/nodro7/angular-mydatepicker)
  - Testing if button click is opening calendar or rather if this component exists in the DOM after button click
- Move Helpbutton to its right module
- Prepare platform check to login page
- Fix other consumption calculation
  - Adding unittest for calculation of other consumption
  - Fixing wrong calculation of other consumption for multiple evcs and consumptionMeters
- Refactor gridOptimizedCharge history
- Replace global role with edge role check
  - /settings: replacing check for global Role with edge Role
  - Add unittest for role in /settings
- Live-EnergyMonitor: use new sum Max/Min ESS DischargePower channels
- refactor variables and methods in right order
  - Refactors every variable and method to member-ordering rule.
  - See: [https://typescript-eslint.io/rules/member-ordering/]

---------

Co-authored-by: Sebastian Asen <47855186+sebastianasen@users.noreply.github.com>
Co-authored-by: Lukas Rieger <73471197+lukasrgr@users.noreply.github.com>
Co-authored-by: Johann Kaufmann <165755282+johannk24@users.noreply.github.com>
Co-authored-by: Michael Grill <59126309+michaelgrill@users.noreply.github.com>
Co-authored-by: Jonas Ernst <11683792+JonasE99@users.noreply.github.com>
Co-authored-by: Stefan Feilmeier <3515268+sfeilmeier@users.noreply.github.com>
Co-authored-by: Hueseyin Sahutoglu <34771592+huseyinsaht@users.noreply.github.com>
Co-authored-by: Fabian Brandtner <10850256+fabian94533@users.noreply.github.com>
Co-authored-by: Anas Shetla <141644226+AnasShetla@users.noreply.github.com>
